examples/deepcopy-gen/generators/deepcopy.go

```diff
@@ -286,27 +286,76 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
 	return true
 }
 
-// hasDeepCopyMethod returns true if an appropriate DeepCopy() method is
-// defined for the given type.  This allows more efficient deep copy
+// deepCopyMethod returns the signature of a DeepCopy() method, nil or calls glog.Fatalf
+// if the type does not match. This allows more efficient deep copy
 // implementations to be defined by the type's author.  The correct signature
 // for a type T is:
 //    func (t T) DeepCopy() T
 // or:
-//    func (t *T) DeepCopy() T
-func hasDeepCopyMethod(t *types.Type) bool {
-	for mn, mt := range t.Methods {
-		if mn != "DeepCopy" {
-			continue
-		}
-		if len(mt.Signature.Parameters) != 0 {
-			return false
-		}
-		if len(mt.Signature.Results) != 1 || mt.Signature.Results[0].Name != t.Name {
-			return false
-		}
-		return true
+//    func (t *T) DeepCopy() *T
+func deepCopyMethod(t *types.Type) *types.Signature {
+	f, found := t.Methods["DeepCopy"]
+	if !found {
+		return nil
 	}
-	return false
+	if len(f.Signature.Parameters) != 0 {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected no parameters", t)
```

Comment:
Why are we aborting?  Other tools just ignore signatures that don't match.  e.g.

https://play.golang.org/p/tl05gzSswY-

Reply:
Having a wrong signature is most probably a programmer error. And assuming the custom deepcopy being called, but isn't, sound dangerous and hard to debug. I prefer to fail early here.

Reply:
I second that opinion. Chasing bugs like that, which can be easily caught earlier is definitely better option.